src/cells/model.ts

```diff
@@ -111,153 +75,62 @@ interface IBaseCellModel extends IDisposable {
  */
 export
-interface ICodeCellModel extends IBaseCellModel {
-  /**
-   * Execution, display, or stream outputs.
-   */
-  output: IOutputAreaModel;
-
+interface ICodeCellModel extends ICellModel {
   /**
    * The code cell's prompt number. Will be null if the cell has not been run.
    */
   executionCount: number;
 
   /**
-   * Whether the cell is collapsed/expanded.
-   */
-  collapsed?: boolean;
-
-  /**
-   * Whether the cell's output is scrolled, unscrolled, or autoscrolled.
-   */
-  scrolled?: ScrollSetting;
-
-  /**
-   * Clear the cell state.
-   */
-  clear(): void;
-}
-
-
-/**
- * The definition of a raw cell.
- */
-export
-interface IRawCellModel extends IBaseCellModel {
-  /**
-   * Raw cell metadata format for nbconvert.
-   */
-  format?: string;
-}
-
-
-/**
- * The definition of a markdown cell.
- */
-export
-interface IMarkdownCellModel extends IBaseCellModel {
-  /**
-   * Whether the cell is rendered.
+   * The cell outputs.
    */
-  rendered: boolean;
+  outputs: ObservableOutputs;
 }
 
 
-
-/**
- * A model consisting of any valid cell type.
- */
-export
-type ICellModel =  (
-  IRawCellModel | IMarkdownCellModel | ICodeCellModel
-);
-
-
 /**
- * An implemention of the base cell Model.
+ * An implementation of the cell model.
  */
 export
-class BaseCellModel implements IBaseCellModel {
-  /**
-   * Construct a new base cell model.
-   */
-  constructor(input: IInputAreaModel) {
-    this._input = input;
-  }
-
-  /**
-   * A signal emitted when the state of the model changes.
-   */
-  get stateChanged(): ISignal<IBaseCellModel, IChangedArgs<any>> {
-    return CellModelPrivate.stateChangedSignal.bind(this);
-  }
-
-  /**
-   * A signal emitted when a user metadata state changes.
-   *
-   * #### Notes
-   * The signal argument is the namespace of the metadata that changed.
-   */
-  get metadataChanged(): ISignal<IBaseCellModel, string> {
-    return CellModelPrivate.metadataChangedSignal.bind(this);
-  }
-
-  /**
-   * Get the input area model.
-   */
-  get input(): IInputAreaModel {
-    return this._input;
-  }
-
+class CellModel implements ICellModel {
   /**
-   * The trusted state of the cell.
-   *
-   * See http://jupyter-notebook.readthedocs.org/en/latest/security.html.
+   * Construct a cell model from optional cell content.
    */
-  get trusted(): boolean {
-    return this._trusted;
-  }
-  set trusted(newValue: boolean) {
-    if (newValue === this._trusted) {
+  constructor(cell?: IBaseCell) {
+    if (!cell) {
       return;
     }
-    let oldValue = this._trusted;
-    this._trusted = newValue;
-    this.onTrustChanged(newValue);
-    let name = 'trusted';
-    this.stateChanged.emit({ name, oldValue, newValue });
+    this.source = cell.source;
+    let metadata = utils.copy(cell.metadata);
+    if (this.type !== 'raw') {
+      delete metadata['format'];
+    }
+    if (this.type !== 'code') {
+      delete metadata['collapsed'];
+      delete metadata['scrolled'];
+    }
```

Comment:
This is a bit weird that the base class is doing things that rely on `this.type` and an understanding of how subclasses use metadata. And arguably if a markdown cell happens to have metadata with key 'collapsed', and gets converted to a code cell, the metadata key should now means something and maybe should be reset anyway. I experimented with some other implementations where the base class just copies the metadata and then the subclasses take care of the extra metadata fields, and concluded that the problem here really is about mixing a user metadata bag with metadata fields in the standard. If those were separated, we could copy the user metadata in the base class and in subclasses decide how to deal with cell-type-specific metadata.